src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs

```diff
@@ -752,119 +809,129 @@ private static bool HasApplicableConditionalMethod(OverloadResolutionResult<Meth
+                    {
+                        argumentSyntax = arg.Syntax;
+                        return true;
+                    }
+                }
+            }
+
+            argumentSyntax = null;
+            return false;
+        }
+
+        private BoundExpression TryEarlyBindSingleCandidateInvocationWithDynamicArgument(
             SyntaxNode syntax,
             SyntaxNode expression,
             string methodName,
             BoundMethodGroup boundMethodGroup,
             BindingDiagnosticBag diagnostics,
             CSharpSyntaxNode queryClause,
-            MethodGroupResolution resolution)
+            MethodGroupResolution resolution,
+            MemberResolutionResult<MethodSymbol> methodResolutionResult)
         {
-            // Invocations of local functions with dynamic arguments don't need
-            // to be dispatched as dynamic invocations since they cannot be
-            // overloaded. Instead, we'll just emit a standard call with
-            // dynamic implicit conversions for any dynamic arguments. There
-            // are two exceptions: "params", and unconstructed generics. While
-            // implementing those cases with dynamic invocations is possible,
-            // we have decided the implementation complexity is not worth it.
-            // Refer to the comments below for the exact semantics.
-
-            Debug.Assert(resolution.IsLocalFunctionInvocation);
-            Debug.Assert(resolution.OverloadResolutionResult.Succeeded);
-            Debug.Assert(queryClause == null);
+            MethodSymbol singleCandidate = methodResolutionResult.LeastOverriddenMember;
+            if (boundMethodGroup.TypeArgumentsOpt.IsDefaultOrEmpty && singleCandidate.IsGenericMethod)
+            {
+                // If we call an unconstructed generic function with a
+                // dynamic argument in a place where it influences the type
+                // parameters, we need to dynamically dispatch the call (as the
+                // function must be constructed at runtime). We disallow that
+                // when we know that runtime binder will not be able to handle the case.
+                // See https://github.com/dotnet/roslyn/issues/21317
 
-            var validResult = resolution.OverloadResolutionResult.ValidResult;
-            var args = resolution.AnalyzedArguments.Arguments.ToImmutable();
-            var refKindsArray = resolution.AnalyzedArguments.RefKinds.ToImmutableOrNull();
+                // However, doing a specific analysis of each
+                // argument and its corresponding parameter to check if it's
+                // generic (and allow dynamic in non-generic parameters) doesn't
+                // seem to worth the complexity. So, just disallow any mixing of dynamic and
+                // inferred generics. (Explicit generic arguments are fine)
 
-            ReportBadDynamicArguments(syntax, args, refKindsArray, diagnostics, queryClause);
```

Comment:
>ReportBadDynamicArguments(syntax, args, refKindsArray, diagnostics, queryClause); [](http://example.com/codeflow?start=12&length=81)

Note, I intentionally removed this check, because I believe the spec doesn't require to perform it for statically bound invocations.

Reply:
For example, the helper implements "It is a compile time error if a method invocation is dynamically bound and any of the parameters, including the receiver, has the in modifier." rule from [https://github.com/dotnet/csharpstandard/blob/draft-v9/standard/expressions.md#1231-general](https://github.com/dotnet/csharpstandard/blob/draft-v9/standard/expressions.md#1231-general)